diff --git a/include/ds/lce_classic_for_sss.hpp b/include/ds/lce_classic_for_sss.hpp
index 483d9d5..a50ce78 100644
--- a/include/ds/lce_classic_for_sss.hpp
+++ b/include/ds/lce_classic_for_sss.hpp
@@ -141,6 +141,17 @@ class lce_classic_for_sss {
 #endif
 #endif
 
+      m_sa = lce::util::bit_aligned_vector(sa.size(), reduced_size);
+
+#pragma omp parallel
+      {
+        const uint64_t threads = omp_get_num_threads();
+        const uint64_t chunk = (sa.size() + threads - 1) / threads;
+        const uint64_t beg = std::min<uint64_t>(sa.size(), omp_get_thread_num() * chunk);
+        const uint64_t end = std::min<uint64_t>(sa.size(), beg + chunk);
+        m_sa.fill(beg, end, [&](uint64_t i) { return uint64_t(int64_t(sa[i])); });
+      }
+
       sa = std::decay_t<decltype(sa)>();
 
       uint64_t max_lcp = 0;
@@ -226,9 +237,16 @@ class lce_classic_for_sss {
     return m_lcp[m_rmq.rmq_shifted(m_isa[l], m_isa[r])];
   }
 
+  const lce::util::bit_aligned_vector& get_sa_s() const { return m_sa; }
+
+  const lce::util::bit_aligned_vector& get_isa_s() const { return m_isa; }
+
+  void free_sa_s() { m_sa = lce::util::bit_aligned_vector(); }
+
  private:
   size_t m_size;
   uint64_t m_tau;
+  lce::util::bit_aligned_vector m_sa;
   lce::util::bit_aligned_vector m_isa;
   lce::util::bit_aligned_vector m_lcp;
   lce::rmq::rmq_n<t_index_type, uint32_t, 128, lce::util::bit_aligned_view> m_rmq;
diff --git a/include/ds/lce_sss.hpp b/include/ds/lce_sss.hpp
index 7301dd1..1aff567 100644
--- a/include/ds/lce_sss.hpp
+++ b/include/ds/lce_sss.hpp
@@ -231,6 +231,12 @@ class lce_sss {
 
   const lce::util::bit_aligned_vector& get_sync_set() const { return m_sync_set.get_sss(); }
 
+  const lce::util::bit_aligned_vector& get_sa_s() const { return m_sss_lce.get_sa_s(); }
+
+  const lce::util::bit_aligned_vector& get_isa_s() const { return m_sss_lce.get_isa_s(); }
+
+  void free_sa_s() { m_sss_lce.free_sa_s(); }
+
  private:
   text_type m_text;
   size_t m_size;
