diff --git a/malloc_count.c b/malloc_count.c
index a1deac1..ac0bbdf 100644
--- a/malloc_count.c
+++ b/malloc_count.c
@@ -38,10 +38,10 @@
 /* user-defined options for output malloc()/free() operations to stderr */
 
 static const int log_operations = 0;    /* <-- set this to 1 for log output */
-static const size_t log_operations_threshold = 1024*1024;
+static const size_t log_operations_threshold = 1024*1024*2;
 
 /* option to use gcc's intrinsics to do thread-safe statistics operations */
-#define THREAD_SAFE_GCC_INTRINSICS      0
+#define THREAD_SAFE_GCC_INTRINSICS      1
 
 /* to each allocation additional data is added for bookkeeping. due to
  * alignment requirements, we can optionally add more than just one integer.
@@ -64,6 +64,10 @@ static aligned_alloc_type real_aligned_alloc = NULL;
 static free_type real_free = NULL;
 static realloc_type real_realloc = NULL;
 
+static void init(void);
+
+static int mc_initializing = 0;
+
 /* a sentinel value prefixed to each allocation */
 static const size_t sentinel = 0xDEADC0DE;
 
@@ -162,6 +166,12 @@ extern void* malloc(size_t size)
 
     if (size == 0) return NULL;
 
+    if (!real_malloc && !mc_initializing) {
+        mc_initializing = 1;
+        init();
+        mc_initializing = 0;
+    }
+
     if (real_malloc)
     {
         /* call real malloc procedure in libc */
@@ -393,6 +403,8 @@ static __attribute__((constructor)) void init(void)
 {
     char *error;
 
+    if (real_malloc) return;
+
     setlocale(LC_NUMERIC, ""); /* for better readable numbers */
 
     dlerror();
@@ -424,9 +436,6 @@ static __attribute__((constructor)) void init(void)
 
 static __attribute__((destructor)) void finish(void)
 {
-    fprintf(stderr, PPREFIX
-            "exiting, total: %'lld, peak: %'lld, current: %'lld\n",
-            total, peak, curr);
 }
 
 /*****************************************************************************/
